# Форматирование строк

Форматирование строк позволяет Вам добавлять элементы внутрь строки вместо того, чтобы соединять строки вместе с помощью запятых или конкатенации. В качестве примера, сравните:

    player = 'Thomas'
    points = 33
    
    'Last night, '+player+' scored '+str(points)+' points.'  # конкатенация
    
    f'Last night, {player} scored {points} points.'          # форматирование строк


Форматирование строк можно выполнить тремя способами.
* Старый способ - это использование %s, которые потом заменяются на значения.
* Улучшенный способ использует метод `.format()`.
* Новый способ, появившийся в Python 3.6: f-строки (*f-strings*).

Поскольку Вы можете встретить все три варианта в коде, который пишут другие люди, то давайте разберём все три варианта.

## Форматирование с помощью замены %s
Можно использовать <code>%s</code>, чтобы вставить строку внутрь операции print. Символ `%` в этом случае называется оператором форматирования строки ("string formatting operator").

In [1]:
print("I'm going to inject %s here." %'something')

I'm going to inject something here.


Вы можете передать несколько значений, для этого нужно поместить их после оператора `%` в виде кортежа (tuple). Очень скоро у нас будет лекция про кортежи, но в двух словах - это несколько значений, указанные в круглых скобках через запятую.

In [2]:
print("I'm going to inject %s text here, and %s text here." %('some','more'))

I'm going to inject some text here, and more text here.


Вы также можете указывать имена переменных:

In [3]:
x, y = 'some', 'more'
print("I'm going to inject %s text here, and %s text here."%(x,y))

I'm going to inject some text here, and more text here.


### Методы конвертации.
Два метода - <code>%s</code> и <code>%r</code> - могут конвертировать любой объект python в строку, используя два различных метода: `str()` and `repr()`. Мы изучим эти функции более подобно позднее, но сразу скажем, что `%r` и `repr()` представляют строковое представление объекта (*string representation*), включая кавычки и любые символы экранирования.

In [4]:
print('He said his name was %s.' %'Fred')
print('He said his name was %r.' %'Fred')

He said his name was Fred.
He said his name was 'Fred'.


Ещё один пример: `\t` добавляет табуляцию.

In [5]:
print('I once caught a fish %s.' %'this \tbig')
print('I once caught a fish %r.' %'this \tbig')

I once caught a fish this 	big.
I once caught a fish 'this \tbig'.


Оператор `%s` преобразует всё, что он видит, в строку, включая целые и дробные числа. Оператор `%d` сначала преобразует числа в целые числа integer, без округления. Обратите внимание на разницу:

In [6]:
print('I wrote %s programs today.' %3.75)
print('I wrote %d programs today.' %3.75)   

I wrote 3.75 programs today.
I wrote 3 programs today.


### Числа с плавающей точкой: точность и добавление пробелов
Числа с плавающей точкой (Floating point numbers) используют следующий формат: <code>%5.2f</code>. Здесь <code>5</code>  - это минимальное количество символов, которое будет содержаться в строке; если всё число не содержит достаточно цифр, то строка будет дополнена пробелами. Далее, <code>.2f</code> указывает, сколько цифр показать после разделителя дробной части. Давайте посмотрим несколько примеров:

In [7]:
print('Floating point numbers: %5.2f' %(13.144))

Floating point numbers: 13.14


In [8]:
print('Floating point numbers: %1.0f' %(13.144))

Floating point numbers: 13


In [9]:
print('Floating point numbers: %1.5f' %(13.144))

Floating point numbers: 13.14400


In [10]:
print('Floating point numbers: %10.2f' %(13.144))

Floating point numbers:      13.14


In [11]:
print('Floating point numbers: %25.2f' %(13.144))

Floating point numbers:                     13.14


Более подробно про форматирование строк с использованием % можно почитать в документации вот здесь: https://docs.python.org/3/library/stdtypes.html#old-string-formatting

### Сочетание различных видов форматирования
В одной команде вполне можно использовать более одного способа форматирования:

In [12]:
print('First: %s, Second: %5.2f, Third: %r' %('hi!',3.1415,'bye!'))

First: hi!, Second:  3.14, Third: 'bye!'


## Форматирование с помощью метода `.format()` 
Чтобы выполнить форматирование объектов в строки для команды print, лучше использовать метод `.format()`. Синтаксис следующий:

    'String here {} then also {}'.format('something1','something2')
    
Например:

In [13]:
print('This is a string with an {}'.format('insert'))

This is a string with an insert


### Метод .format() имеет несколько преимуществ по сравнению с методом %s:

#### 1. Добавляемые объекты можно указать по их номеру:

In [14]:
print('The {2} {1} {0}'.format('fox','brown','quick'))

The quick brown fox


#### 2. Добавляемым объектам можно присвоить названия:

In [15]:
print('First Object: {a}, Second Object: {b}, Third Object: {c}'.format(a=1,b='Two',c=12.3))

First Object: 1, Second Object: Two, Third Object: 12.3


#### 3. Добавляемые объекты можно переиспользовать, без дублирования:

In [16]:
print('A %s saved is a %s earned.' %('penny','penny'))
# vs.
print('A {p} saved is a {p} earned.'.format(p='penny'))

A penny saved is a penny earned.
A penny saved is a penny earned.


### Выравнивание, добавление пробелов и точность для `.format()`
Внутри фигурных скобок можно указать длину полей, выравнивание влево/вправо, параметры округления и т.д.

In [17]:
print('{0:8} | {1:9}'.format('Fruit', 'Quantity'))
print('{0:8} | {1:9}'.format('Apples', 3.))
print('{0:8} | {1:9}'.format('Oranges', 10))

Fruit    | Quantity 
Apples   |       3.0
Oranges  |        10


По умолчанию, `.format()` выравнивает текст влево, а числа вправо. Вы можете указать опциональные символы  `<`,`^` или `>`, чтобы установить выравнивание влево, по центру или вправо:

In [18]:
print('{0:<8} | {1:^8} | {2:>8}'.format('Left','Center','Right'))
print('{0:<8} | {1:^8} | {2:>8}'.format(11,22,33))

Left     |  Center  |    Right
11       |    22    |       33


Перед оператором выравнивания можно указать символ, с помощью которого строка будет дополняться до нужной длины.

In [19]:
print('{0:=<8} | {1:-^8} | {2:.>8}'.format('Left','Center','Right'))
print('{0:=<8} | {1:-^8} | {2:.>8}'.format(11,22,33))

Left==== | -Center- | ...Right
11====== | ---22--- | ......33


Ширина полей и точность дробных чисел указываются похоже на синтаксис %. Следующие две команды print эквивалентны:

In [20]:
print('This is my ten-character, two-decimal number:%10.2f' %13.579)
print('This is my ten-character, two-decimal number:{0:10.2f}'.format(13.579))

This is my ten-character, two-decimal number:     13.58
This is my ten-character, two-decimal number:     13.58


Обратите внимание, что после двоеточия идут 5 пробелов, и 5 символов занимает число 13.58 - всего получается 10 символов.

Более подробно о форматировании строк с помощью метода `.format()` можно почитать в документации вот здесь: https://docs.python.org/3/library/string.html#formatstrings

## f-строки (f-strings)

Добавленные в Python 3.6, f-строки имеют несколько преимуществ по сравнению с методом `.format()`, который мы рассмотрели выше. Одно из преимуществ - Вы можете указать внешние переменные напрямую в строке, а не передавать их как параметры через  `.format(var)`.

In [21]:
name = 'Fred'

print(f"He said his name is {name}.")

He said his name is Fred.


Укажите `!r`, чтобы получить строковое представление:

In [22]:
print(f"He said his name is {name!r}")

He said his name is 'Fred'


#### Формат для дробных чисел следующий: `"result: {value:{width}.{precision}}"`

Если для метода `.format()` Вы можете увидеть `{value:10.4f}`, то для f-строк этот формат становится вот таким: `{value:{10}.{6}}`


In [23]:
num = 23.45678
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:{10}.{6}}")

My 10 character, four decimal number is:   23.4568
My 10 character, four decimal number is:   23.4568


Обратите внимание, что для f-строк точность *precision* означает общее количество цифр, а не только цифры после разделителя дробной части. Это лучше соответствует научной нотации и статистическому анализу. К сожалению, f-строки не дополняются нулями справа от дробной части, даже если точность это позволяет:

In [24]:
num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:{10}.{6}}")

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:     23.45


Если это для Вас будет важно, Вы всегда можете использовать метод `.format()` внутри f-string:

In [25]:
num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:10.4f}")

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:   23.4500


Более подробно о форматированных строковых литералах можно почитать в документации вот здесь: https://docs.python.org/3/reference/lexical_analysis.html#f-strings

Это были основы форматирования строк!